_drafts/faster-builds.md

```diff
@@ -0,0 +1,649 @@
+2. For headers that are included frequently and/or include many other headers,
+   consider using forward declarations[^11]. For instance, let's look at the
+   following header file.
+
+   ```c
+   // settings_file.h
+
+   // Very Large Header File!
+   #include "filesystem.h"
+
+   bool settings_file_open(File *file);
+   ```
+
+   We are including `filesystem.h` _only_ to have an opaque pointer to a `File`
+   defined, but we are paying the price by including such a large file. Since we
+   don't need the full internal definition of `File`, we can instead use a
+   forward declaration.
+
+   ```c
+   // settings_file.h
+
+   typedef struct File File;
+
+   bool settings_file_open(File *file);
+   ```
+
+   If you find yourself needing to forward declare this more than once, I
+   suggest creating a header file `filesystem_types.h` which contains these
+   declarations. Now other files can include this new `_types.h` header instead.
+
+   ```c
+   // filesystem_types.h
+
+   typedef struct File File;
+
+   // Other forward declarations
+   ```
+
+3. Use [include guards](https://en.wikipedia.org/wiki/Include_guard). I suggest
+   using `#pragma once` since it is less error prone and is quite portable[^14].
+   The standard C way works too.
+4. Wrap third party library code in an abstraction layer so its internal types
+   and headers do not leak where they do not belong. Vendors are not in the
+   business of clean and quick compiling code, and you shouldn't be at the mercy
+   of their bad habits.
+
+On a previous project, the most significant patch I made related to header
+dependencies resulted in a compilation speedup of 30%. Before the change, almost
+every file in our firmware project had incidentally included the entire set of
+our vendor's peripheral library header files, which are usually **huge**. Each
+`.c` file was accidentally including 600+ headers.
+
+The fix was to create a couple of `..._types.h` files which contained forward
+declarations to help break the dependencies between our code and vendor code.
+
+### Pre-compiled headers
+
+> NOTE: This is a GCC/Clang only feature.
+
+After cleaning up all of the header dependency issues that you can find, there
+may still be some headers that you can't avoid including everywhere, and they
+are slowing down your build. The good news is that there is a way to pre-compile
+your headers with GCC[^12].
+
+Pre-compiling a header means that the compiler will only need to compile a
+header _once_, and whenever it is included by other files, it will reuse the
+built object. It compiles the header into an intermediate file with the
+extension `.h.gch`. If the location of these files is added as an include path
+when compiling (before the real headers), GCC will use the pre-compiled versions
+instead.
+
+Let's go over the easiest way to do this in GCC, one that doesn't require you to
+re-architect your project.
+
+1. Create a header file `precompiled.h` and place it in your source tree.
```

Comment:
If possible, try to avoid the numbered list and code snippets for more visual zen.

> Create a header file precompiled.h and place it in your source tree. Add a rule to compile this header, just as you would with a `.c` file. It should be built before building the `.c` or `.cpp` files. Then include this file using a CFLAG argument `-include <>` when compiling the `.c` files. With Make, this looks something like:
>
> ```
> PRECOMPILED_H := "precompiled.h"
> PRECOMPILED_H_OBJ := $(PRECOMPILED_H).gch
> 
> %.h.gch: $(PRECOMPILED_H)
> 	$(CC) $(CFLAGS) -c -o $@ $<
>
> %.o : %.c | $(PRECOMPILED_H_OBJ)
> 	$(CC) $(CFLAGS) -include $(PRECOMPILED_H) -c -o $@ $<
>  ```